winml/lib/Api.Core/WinMLAdapter.cpp

```diff
@@ -0,0 +1,840 @@
+        auto alloc = registrations[onnxruntime::CPU].factory(0);
+       
+       onnxruntime::TensorShape tensor_shape(*shape);
+
+        // Unowned raw tensor pointer passed to engine
+        auto tensor = new onnxruntime::Tensor(
+            data_type,
+            tensor_shape,
+            buffer,
+            alloc->Info());
+
+        auto ort_value = wil::MakeOrThrow<AbiSafeOrtValue>();
+        ort_value->get().Init(tensor,
+            onnxruntime::DataTypeImpl::GetType<onnxruntime::Tensor>(),
+            onnxruntime::DataTypeImpl::GetType<onnxruntime::Tensor>()->GetDeleteFunc());
+
+        *cpu_value = ort_value.Detach();
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE CreateMLValue(
+        winml::TensorKind kind,
+        onnxruntime::MLDataType data_type,
+        const int64_t * shape,
+        uint32_t shape_count,
+        onnxruntime::IExecutionProvider* provider,
+        IOrtValue ** ort_value) override {
+        onnxruntime::TensorShape tensor_shape(shape, shape_count);
+        auto tensor = new onnxruntime::Tensor(
+            GetType(kind),
+            tensor_shape,
+            provider->GetAllocator(0, ::OrtMemType::OrtMemTypeDefault));
+        auto ort_value_out = wil::MakeOrThrow<AbiSafeOrtValue>();
+        ort_value_out->get().Init(tensor,
+            data_type,
+            data_type->GetDeleteFunc());
+
+        *ort_value = ort_value_out.Detach();;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE CreateOrtValue(
+        void * data,
+        onnxruntime::MLDataType data_type,
+        IOrtValue ** ort_value) override {
+
+        auto ort_value_out = wil::MakeOrThrow<AbiSafeOrtValue>();
+        ort_value_out->get().Init(
+            data,
+            data_type,
+            data_type->GetDeleteFunc());
+
+        *ort_value = ort_value_out.Detach();
+        return S_OK;
+    }
+
+    // Override select shape inference functions which are incomplete in ONNX with versions that are complete,
+    // and are also used in DML kernel registrations.  Doing this avoids kernel and shader creation being
+    // deferred until first evaluation.  It also prevents a situation where inference functions in externally
+    // registered schema are reachable only after upstream schema have been revised in a later OS release,
+    // which would be a compatibility risk.
+    HRESULT STDMETHODCALLTYPE OverrideSchemaInferenceFunctions() override {
+        static std::once_flag schema_override_once_flag;
+        std::call_once(schema_override_once_flag, []() {
+            SchemaInferenceOverrider::OverrideSchemaInferenceFunctions();
+        });
+        return S_OK;
+    }
+
+
+};
+
+extern "C"
+HRESULT STDMETHODCALLTYPE OrtGetWinMLAdapter(IWinMLAdapter** adapter) {
+    // make an adapter instance
+    Microsoft::WRL::ComPtr<WinMLAdapter> adapterptr = wil::MakeOrThrow<WinMLAdapter>();
+    return adapterptr.CopyTo(__uuidof(IWinMLAdapter), (void **)adapter);
```

Comment:
>(void **) [](start = 54, length = 9)

static_cast?

Reply:
static_cast doesn't work to void ** .  fun, right?

[https://stackoverflow.com/questions/3625410/c-static-cast-from-float-to-void](https://stackoverflow.com/questions/3625410/c-static-cast-from-float-to-void)


---
In reply to: [347044742](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347044742) [](ancestors = 347044742)

Reply:
reinterpret! that's what I meant. my bad.

---
In reply to: [347542211](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347542211) [](ancestors = 347542211,347044742)